gnd/src/main/java/com/google/android/gnd/system/auth/AuthenticationManagerImpl.java

```diff
@@ -45,10 +44,9 @@
 import java8.util.Optional;
 import javax.inject.Inject;
 
-@ActivityScoped
-public class AuthenticationManager {
-  private static final String TAG = AuthenticationManager.class.getSimpleName();
-  private static final int SIGN_IN_REQUEST_CODE = AuthenticationManager.class.hashCode() & 0xffff;
+public class AuthenticationManagerImpl implements AuthenticationManager {
```

Comment:
As mentioned by VC, could Spies be helpful here? In that case you can mock out some methods while using the actual implementation of others.
https://github.com/mockito/mockito/wiki/Using-Spies-(and-Fakes)

I realize spies aren't a silver bullet; there will be cases where creating an interface + impl + testImpl will be cleaner, but for simple cases perhaps it's worth considering?

Reply:
I've investigated using spies and have come to the conclusion that coding to an interface (aka the API approach) is still the best way to proceed. Here's why: 

- We're already using it. For example, we have a `RemoteStorageManager` interface which is implemented by `FirestoreStorageManager`, and a `RemoteDataStore` which is implemented by `FirestoreDataStore`. 
- When there is more than one implementation of a given interface (e.g. one for prod, one for test) we should reflect that in the code, rather than relying on a spy to do that work for us.  
- Speaking to various people with more testing experience than me, the general feeling around spies is that they are a code smell. There's also a shift away from using mocks. 
- It makes it easy to add new authentication methods, such as Twitter, Facebook, email address. Potentially we might add some anonymous login which can then be converted to an account if the user wishes later. 

Incidentally, the `Impl` suffix adds noise ([and is an anti-pattern](https://octoperf.com/blog/2016/10/27/impl-classes-are-evil/)) we can avoid this by naming the implementations according to their specialization e.g. `GoogleAuthenticationManager` (I'll commit a change for this shortly). 

WDYT?

Reply:
This makes sense. Int his case, I assume won't need interface/impls for all classes, only on those classes that touch remote or system services so that we can reliably and quickly run UI tests? If so, I think this amount of additional complexity is a reasonable price to pay for fast, reproducible UI tests. Please go ahead!